src/super_gradients/training/transforms/processing.py

```diff
@@ -0,0 +1,186 @@
+from typing import Tuple, List, Union
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import numpy as np
+
+from super_gradients.training.transforms.utils import (
+    _rescale_image,
+    _rescale_bboxes,
+    _pad_image_on_side,
+    _get_center_padding_params,
+    _shift_image,
+    _shift_bboxes,
+)
+
+
+@dataclass
+class ProcessingMetadata(ABC):
+    """Metadata including information to postprocess a prediction."""
+
+
+@dataclass
+class ComposeProcessingMetadata(ProcessingMetadata):
+    metadata_lst: List[Union[None, ProcessingMetadata]]
+
+
+@dataclass
+class DetectionPadToSizeMetadata(ProcessingMetadata):
+    shift_w: float
+    shift_h: float
+
+
+@dataclass
+class RescaleMetadata(ProcessingMetadata):
+    original_size: Tuple[int, int]
+    sy: float
+    sx: float
+
+
+class Processing(ABC):
+    @abstractmethod
+    def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, Union[None, ProcessingMetadata]]:
+        """Processing an image, before feeding it to the network. Expected to be in (H, W, C) or (H, W)."""
+        pass
+
+    @abstractmethod
+    def postprocess_predictions(self, predictions: np.ndarray, metadata: Union[None, ProcessingMetadata]) -> np.ndarray:
```

Comment:
predictions is [N,4] or [N,4+]? what is the format of bboxes here?
I suggest to require exactly [N,4] of specific format. Since we apply spatial transformations only on box coordinates not on labels and scores there is no need to propagate this information.
So we probably can simplify the API by explicitly saying `predictions` -> `bboxes` has shape of [N,4] of xyxy format.

Reply:
I agree that this function doesn't need to know about the whole prediction, that's a good point.
The thing is that it moves the responsibility to the processing pipeline to handle this (to know how to get bbox, label, ...).
The current implementation of the pipeline is (almost) fully task agnostic which I believe is the right way of doing, and to do so it relies on the processing functions to handle the predictions themselves. It also relies on the postprocessing step to return targets in xyxy_score_label format.


The better solution I can think of is making predictions an instance of a class that would abstract away how to get the bbox, score, label. This would move away the responsibility of knowing how to get the target/score/label away from both pipeline and the processing function.
This abstraction can be useful in different contexts but are we sure we want it ? In particular, this would impact the functions that are in utils which are also being used in `transforms.py`, so it looks like it would be a big refacto with a risk of breaking change to handle.

Reply:
💯 `DetectionPrediction` ❤️ ❤️ ❤️

Reply:
Something like this ? I just wrote it quickly, didnt test and didnt add it to the code
```

@dataclass
class DetectionPredictions(Prediction):
    bboxes: np.ndarray
    confidence: np.ndarray
    labels: np.ndarray

    bbox_format: str
    image_shape: Tuple[int, int]

    @property
    def bboxes_cxcywh(self):
        return self._get_bbox_as("cxcywh")

    @bboxes_cxcywh.setter
    def bboxes_cxcywh(self, bboxes: np.ndarray):
        self._set_bbox_from(bboxes=bboxes, input_bbox_format="cxcywh")

    @property
    def bboxes_xywh(self):
        return self._get_bbox_as("xywh")

    @bboxes_xywh.setter
    def bboxes_xywh(self, bboxes: np.ndarray):
        self._set_bbox_from(bboxes=bboxes, input_bbox_format="xywh")

    @property
    def bboxes_normalized_xyxy(self):
        return self._get_bbox_as("normalized_xyxy")

    @bboxes_normalized_xyxy.setter
    def bboxes_normalized_xyxy(self, bboxes: np.ndarray):
        self._set_bbox_from(bboxes=bboxes, input_bbox_format="normalized_xyxy")

    def _get_bbox_as(self, desired_bbox_format: str):
        return convert_bboxes(
            bboxes=self.bboxes,
            image_shape=self.image_shape,
            source_format=factory.get(self.bbox_format),
            target_format=factory.get(desired_bbox_format),
            inplace=False,
        )

    def _set_bbox_from(self, bboxes: np.ndarray, input_bbox_format: str):
        self.bboxes = convert_bboxes(
            bboxes=bboxes,
            image_shape=self.image_shape,
            source_format=factory.get(input_bbox_format),
            target_format=factory.get(self.bbox_format),
            inplace=False,
        )

    def _set_default_bbox_format(self, desired_format: str):
        self.bbox_format = desired_format
        self.bboxes = self._get_bbox_as(desired_format)

```

Reply:
@BloodAxe This is what I propose, what do you think ? b012d46148801cdbf2d835fec4e3f47abb4482aa